### ISSUE TITLE
Conda Build Templates

### ISSUE THREAD

**Issue by srossross** [conda/conda-build#1095]
Perhaps it would be good to revamp the conda build template files. And make them a first class citizen:

Some ideas:
1. Load templates from `meta.yaml.template` this should result in a auto generated `meta.yaml`
   - This would result in better understanding of failed builds becuse of unclear template variables
2. Currently there are two passes of the  build:  1 - to fetch the build sources 2. to render the meta.yaml from the template. Perhaps we can make this more clear in the yaml file maybe using the `---` yaml syntax
   
   ``` yaml
   # File: meta.yaml.template
   # Required information
   # This can not be templated
   source: ...
   --- 
   # Template vars
   build: 
    number: {{ var }}
   ```
3. Allow for user contributed build extensions (conda packages) that can define template variables etc. eg:
   
   ```
   # File: meta.yaml.template
   source: ...
   build_ext_modules:
     - conda-build-etx-auto-build-numbers
   ---
   build:
      number: {{ my_great_build_number_generator() }}
   ```
   
   in this example `conda build .` would install/require the conda package `conda-build-etx-auto-build-numbers` to be installed into the root environment before the build starts.

cc @kalefranz @mcg1969

**Comment by srossross** [comment]
I started thinking about item 3 when writing a response for Anaconda-Platform/anaconda-server#258

**Comment by srossross** [comment]
cc @msarahan

**Comment by msarahan** [comment]
I don't think you need conda-build-ext-modules.  I think Jinja2's function definition as filters (http://jinja.pocoo.org/docs/dev/api/#writing-filters) will suffice, and that's something we're working on in #966.  There's a nice demo at https://github.com/conda/conda-build/pull/848#issuecomment-212641321

That's number 2 on my priorities, after #953.  I think it will probably take me another couple of weeks to get to it.

**Comment by srossross** [comment]
Thanks @msarahan That solution looks great. 

It would also be nice to have reusable extension components, rather than copy the `jinja_context.py` file from recipe to recipe.

**Comment by msarahan** [comment]
My hope is that there will be 2 jinja_context.py files: one global, and another per-recipe.  Same with the config yaml files.  We'll merge these, with the local stuff overriding the global stuff.  Conda-build will probably maintain a sane base file - this might be a third, which would be overridden by both the user's global, and the recipe-local.
